library/include/ck/library/reference_tensor_operation/cpu/reference_batchnorm_backward_nhwc_c.hpp

```diff
@@ -0,0 +1,305 @@
+                    invVar               = type_convert<AccDataType>(1.0f) /
+                             std::sqrt(type_convert<AccDataType>(arg.epsilon_) + variance);
+                };
+
+                AccDataType bnBiasDiff = type_convert<AccDataType>(0.0f); // Sum on NHW of dy
+                AccDataType bnScaleDiff =
+                    type_convert<AccDataType>(0.0f); // Sum on NHW of dy * norm_x
+
+                // 1) calculate dy * (x - mean) * invVariance
+                // 2) calculate Sum on NHWC of dy
+                // 3) calculate Sum on NHWC of dy * norm_x
+                for(index_t iN = 0; iN < arg.n_; iN++)
+                {
+                    index_t offset_N = iN * arg.h_ * arg.w_ * arg.c_;
+                    for(index_t iH = 0; iH < arg.h_; iH++)
+                    {
+                        index_t offset_H = iH * arg.w_ * arg.c_;
+                        for(index_t iW = 0; iW < arg.w_; iW++)
+                        {
+                            index_t offset_W = iW * arg.c_;
+
+                            auto offset = offset_N + offset_H + offset_W + offset_C;
+
+                            AccDataType x = type_convert<AccDataType>(arg.p_x_[offset]);
+
+                            AccDataType norm_x = (x - mean) * invVar;
+                            AccDataType dy     = type_convert<AccDataType>(arg.p_dy_[offset]);
+
+                            bnBiasDiff += dy;
+                            bnScaleDiff += norm_x * dy;
+                        };
+                    }
+                };
+
+                arg.p_scaleDiff_[offset_C] = type_convert<ScaleDataType>(bnScaleDiff);
+                arg.p_biasDiff_[offset_C]  = type_convert<BiasDataType>(bnBiasDiff);
+
+                // 1) calculate tmp = scaleDiff * (x - mean) * invVariance
+                // 2) calculate dx = 1/nhw * invVariance * scale * (nhw * dy - biasDiff - tmp)
+                for(index_t iN = 0; iN < arg.n_; iN++)
+                {
+                    index_t offset_N = iN * arg.h_ * arg.w_ * arg.c_;
+                    for(index_t iH = 0; iH < arg.h_; iH++)
+                    {
+                        index_t offset_H = iH * arg.w_ * arg.c_;
+                        for(index_t iW = 0; iW < arg.w_; iW++)
+                        {
+                            index_t offset_W = iW * arg.c_;
+
+                            auto offset = offset_N + offset_H + offset_W + offset_C;
+
+                            AccDataType x = type_convert<AccDataType>(arg.p_x_[offset]);
+
+                            AccDataType norm_x = (x - mean) * invVar;
+                            AccDataType dy     = type_convert<AccDataType>(arg.p_dy_[offset]);
+                            AccDataType scale  = type_convert<AccDataType>(arg.p_scale_[offset_C]);
+
+                            AccDataType tmpVal = norm_x * bnScaleDiff;
+
+                            AccDataType dx = type_convert<AccDataType>(1.0f) / reduceSize * invVar *
```

Comment:
Move type_convert<AccDataType>(1.0f) / reduceSize * invVar outside of the loop

Reply:
Common expression is moved out of the loops by [9d9c9](https://github.com/ROCmSoftwarePlatform/composable_kernel/pull/461/commits/f4d67cf8f5595b696a963c6064f62030b269d9c9)

Reply:
remain unchange

Reply:
Very strange!  I could not see the issue after refreshing, should have been solved

Reply:
Can you see the renaming change in [5c08e](https://github.com/ROCmSoftwarePlatform/composable_kernel/pull/461/commits/bb220a7a508e0194cf3585bd5ea944afeac5c08e)

Reply:
Common expression is moved out of the loop by [670b2b](https://github.com/ROCmSoftwarePlatform/composable_kernel/pull/461/commits/3ac68bcc9f661cf7b048099c259ba5f85e670b2b)